modvendor/modvendor.go

```diff
@@ -0,0 +1,223 @@
+		return "", err
+	}
+
+	// TODO(katcipis): remove .git before moving.
+
+	if err := os.MkdirAll(filepath.Dir(clonedir), 0775); err != nil {
+		return "", errors.E(err, "creating mod dir inside vendor")
+	}
+
+	logger.Trace().Msg("moving cloned mod from workdir to clonedir")
+	if err := os.Rename(workdir, clonedir); err != nil {
+		// This may leave intermediary created dirs hanging on vendordir
+		// since we just create all and then delete clone dir on a failure to move.
+		// If we get a lot of errors from os.Rename we may need to handle this
+		// more gracefully, here we assume that os.Rename errors are rare since both
+		// dirs where just created.
+		errs := errors.L()
+		errs.Append(errors.E(err, "moving cloned module"))
+		errs.Append(os.Remove(clonedir))
+		return "", errs.AsError()
+	}
+	return clonedir, nil
+}
+
+// ParseSource parses the given modsource string.
+// The modsource must be a valid Terraform Git/Github source reference as documented in:
+//
+// - https://www.terraform.io/language/modules/sources
+//
+// Source references that are not Git/Github are not supported.
+func ParseSource(modsource string) (Source, error) {
+	ref := ""
+	splitParams := strings.Split(modsource, "?")
+	modsource = splitParams[0]
+
+	if len(splitParams) > 1 {
+		if len(splitParams) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "unexpected extra '?' on source")
+		}
+
+		refParam := splitParams[1]
+		splitRefParam := strings.Split(refParam, "=")
+
+		if len(splitRefParam) != 2 {
+			return Source{}, errors.E(ErrInvalidModSrc, "parsing ref param %q", refParam)
+		}
+
+		if splitRefParam[0] != "ref" {
+			return Source{}, errors.E(ErrInvalidModSrc, "unknown param %q", splitRefParam[0])
+		}
+
+		ref = splitRefParam[1]
+		if ref == "" {
+			return Source{}, errors.E(ErrInvalidModSrc, "ref param %q is empty", refParam)
+		}
+	}
+
+	switch {
+	case strings.HasPrefix(modsource, "github.com"):
+		return Source{
+			URL:  fmt.Sprintf("https://%s.git", modsource),
+			Path: modsource,
+			Ref:  ref,
+		}, nil
+	case strings.HasPrefix(modsource, "git@"):
+		// In git it could be any user@host, but here we are supporting
+		// the specific options allowed by Terraform on module.source:
+		// - https://www.terraform.io/language/modules/sources#github
+		// In this case being Github ssh access.
+		return Source{
+			URL:  modsource,
+			Path: parseGithubAtPath(modsource),
+			Ref:  ref,
+		}, nil
+	case strings.HasPrefix(modsource, "git::"):
+		// Generic git: https://www.terraform.io/language/modules/sources#generic-git-repository
+		path, err := parseGitColonPath(modsource)
+		if err != nil {
+			return Source{}, err
+		}
+		return Source{
+			URL:  strings.TrimPrefix(modsource, "git::"),
+			Path: path,
+			Ref:  ref,
+		}, nil
+
+	default:
+		return Source{}, errors.E(ErrUnsupportedModSrc)
+	}
+}
+
+func parseGithubAtPath(modsource string) string {
+	path := strings.TrimPrefix(modsource, "git@")
+	path = strings.TrimSuffix(path, ".git")
```

Comment:
this is not going to trim the `.git` if the URL has the form: `git@github.com:org/owner/repo.git?ref=test-branch`

Reply:
The modsource is already stripped of parameter at this point. This works because the tests cover this (we have URLs with refs and the paths are correct).